### PR TITLE
For #884: Replace FindBugs with SpotBugs

### PR DESCRIPTION
**Pull request by paulodamaso**
For #884:
- Added dependencies and `SpotBugsValidation` skeleton `SpotBug` checks
- Added `SpotBugsValidationTests` which tests basic `SpotBugsValidator` behavior
- Added puzzle for resuming implementation

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#957](https://github.com/teamed/qulice/issues/957) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@krzyk[/z](https://www.0crat.com/u/krzyk) everybody who has role `REV` is banned at [#957](https://github.com/teamed/qulice/issues/957); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: qulice-spotbugs/pom.xml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.qulice</groupId>
+        <artifactId>qulice</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>qulice-spotbugs</artifactId>
+    <packaging>jar</packaging>
+    <name>qulice-spotbuugs</name>
```
Comment (krzyk) [inline qulice-spotbugs/pom.xml]: @paulodamaso typo, buugs -> bugs :)

File: qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.List;
+import java.util.logging.Level;
+import org.cactoos.text.JoinedText;
+import org.cactoos.text.ReplacedText;
+import org.cactoos.text.UncheckedText;
+
+/**
+ * Validates source code and compiled binaries with SpotBugs.
+ *
+ * @since 0.17
```
Comment (krzyk) [inline qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java]: @paulodamaso let's make it 0.19

File: qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                this,
+                "No classes at %s, no SpotBugs validation",
+                env.outdir()
+            );
+        }
+    }
+
+    @Override
+    public String name() {
+        return "FindBugs";
```
Comment (krzyk) [inline qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java]: @paulodamaso `SpotBugs`

File: qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                        ",",
+                        env.classpath()
+                    ),
+                    "\\",
+                    "/"
+                )
+            ).asString()
+        );
+        return new VerboseProcess(
+            new ProcessBuilder(args), Level.INFO, Level.INFO
```
Comment (krzyk) [inline qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java]: @paulodamaso have you checked if it is possible to use SpotBugs without executing separate process? It wasn't possible with findbugs, but maybe it is with spotbugs (maybe add a todo to validate that, and later describe the outcome in the javadoc of this class).

File: qulice-spotbugs/src/main/java/com/qulice/spotbugs/package-info.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * SpotBugs validator.
+ *
+ * @since 0.17
```
Comment (krzyk) [inline qulice-spotbugs/src/main/java/com/qulice/spotbugs/package-info.java]: @paulodamaso 0.19

File: qulice-spotbugs/src/test/java/com.qulice.spotbugs/SpotBugsValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ */
+package com.qulice.spotbugs;
+
+import com.qulice.spi.Environment;
+import org.junit.Ignore;
+import org.junit.Test;
+
+/**
+ * Test case for {@link SpotBugsValidator}.
+ * @since 0.3
```
Comment (krzyk) [inline qulice-spotbugs/src/test/java/com.qulice.spotbugs/SpotBugsValidatorTest.java]: @paulodamaso 0.19

File: qulice-spotbugs/src/test/java/com.qulice.spotbugs/package-info.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Tests of SpotBugs validator.
+ *
+ * @since 0.3
```
Comment (krzyk) [inline qulice-spotbugs/src/test/java/com.qulice.spotbugs/package-info.java]: @paulodamaso 0.19

**Comment by paulodamaso** [comment]
@krzyk Take a look, made the corrections

**Comment by krzyk** [comment]
@paulodamaso thanks

**Comment by krzyk** [comment]
@rultor merge

**Comment by krzyk** [comment]
@rultor stop

**Comment by krzyk** [comment]
@paulodamaso qulice fails with:
```
[INFO] PMD: qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java[81-81]: Avoid unused private methods such as 'options()'. (UnusedPrivateMethod)
[INFO] PMD: qulice-spotbugs/src/main/java/com/qulice/spotbugs/SpotBugsValidator.java[73-73]: Avoid unused method parameters such as 'env'. (UnusedFormalParameter)
```

**Comment by paulodamaso** [comment]
@krzyk Fixed, removed from `SpotBUgsValidator` all copied `FindBugs` code. Should have done this way from the start.

**Comment by krzyk** [comment]
@paulodamaso thanks

**Comment by krzyk** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job `gh:teamed/qulice#957` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#957](https://github.com/teamed/qulice/issues/957) is now out of scope

**Pull request merged**